A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,249 @@
+
+## Background
+
+Currently, gRPC does not have any first-class support for weighted round robin load balancing. It can be done in a somewhat cumbersome way with the existing `round_robin` LB policy, by including a given address multiple times in the address list. This mechanism can be used with the legacy grpclb protocol, where the control plane can centrally collect backend metrics from the endpoints and compute weights for load balancing based on those metrics. However, this approach requires a lot of machinery.
+
+As part of the effort to move away from `grpclb`, we will now provide first-class weighted round robin functionality via a new client-side LB policy called `weighted_round_robin`. Instead of depending on centralized backend metric collection, this LB policy will obtain the backend metrics directly from the endpoints using the mechanism described in [gRFC A51][A51], which it will use to compute the endpoint weights.
+
+### Related Proposals:
+* [gRFC A34][A34] (pending)
+* [gRFC A51][A51]
+* [gRFC A52][A52]
+
+
+## Proposal
+
+Introduce a new LB policy `weighted_round_robin`. This policy implements client-side load balancing with direct load reports from backends. Either per-call or periodic out-of-band (OOB) load reporting is active at a time where per-call is the default. The policy is otherwise largely a superset of the existing policy `round_robin`.
+
+### LB Policy Config and Parameters
+
+The `weighted_round_robin` LB policy config will be as follows.
+
+```textproto
+message LoadBalancingConfig {
+  oneof policy {
+    WeightedRoundRobinLbConfig weighted_round_robin = 20 [json_name = "weighted_round_robin"];
+  }
+}
+
+message WeightedRoundRobinLbConfig {
+  // Whether to enable out-of-band utilization reporting collection from
+  // the endpoints.  By default, per-request utilization reporting is used.
+  google.protobuf.BoolValue enable_oob_load_report = 1;
+
+  // Load reporting interval to request from the server.  Note that the
+  // server may not provide reports as frequently as the client requests.
+  // Used only when enable_oob_load_report is true.  Default is 10 seconds.
+  google.protobuf.Duration oob_reporting_period = 2;
+
+  // A given endpoint must report load metrics continuously for at least
+  // this long before the endpoint weight will be used.  This avoids
+  // churn when the set of endpoint addresses changes.  Takes effect
+  // both immediately after we establish a connection to an endpoint and
```

Comment:
I think we cannot tell if the remote end remained online, or rarely if it's the same VM since we disconnected. But the only bad case I can think of is where the set of back end keeps changing fast while they remain healthy, but this sounds unexpected to me.

Reply:
I'm not concerned with the set of backends changing. In that case we'll throw away any data we have for the backend so it'll show up as "new" when it gets re-added.

If the server has MAX_CONNECTION_AGE (or MAX_CONNECTION_IDLE, but that's less useful for round-robin like policies) enabled, then it can disconnect the client periodically. That is used to force clients to do DNS lookups to find more backends, for example. There's a few other rarer cases it could be used.

There's also the concern of cases like "disconnected because keepalive failure." Let's say a server is over 100% utilized and falling over. It may not respond promptly and the connection fails. When the client reconnects the client will use average load instead of avoiding it. It is unlikely a single instance will get in that situation when using WRR. But it is still foreign to me (with round-robin-style LB) to tie state to the particular connection. (With pick-first, there's proxies, so such things are easily appropriate.)

Reply:
I don't know what you mean by grace period. I don't see any existing references to such a period. And it isn't clear to me how providing configuration for such a thing would help. I'm fine revisiting this if there become issues mixing with polling resolvers.

Reply:
Eric, I think you have a good point that MAX_CONNECTION_AGE might not play well here.  If that becomes a problem, users can set the blackout period to 0.  This would admittedly be a trade-off between improving handling of MAX_CONNECTION_AGE and potentially getting bad data when a server first restarts (which is what the blackout period is meant to avoid), but at least there's a knob for it.  And I do agree that this is something we can improve later if it actually becomes a problem in practice.

Reply:
Thanks, I think we are good to close this discussion for now then. Agree with what Mark replied.